Supplemental data public docs (#477)
codatio/codat-docs: docs/using-the-api/supplemental-data/overview.md

---
title: "Supplemental data"
sidebar_label: "Overview"
description: "Customize data types with additional properties not included in Codat's out-of-the box data models"
---

## What is supplemental data?

Codat's supported data types include numerous standardized properties. At the same time, some data sources (e.g. Xero) may contain properties not common across other sources, and therefore not covered by our out-of-the-box data model. You can use supplemental data to extend our data type models to fetch, create, or update such properties alongside our standard ones. 

> Consider our Xero integration as an example. Codat's `suppliers` data type uses Xero's [Contact](https://developer.xero.com/documentation/api/accounting/contacts) schema. Some of its properties (like `TaxNumber`) are mapped by us by default, whereas others (like `BankAccountDetails`) are not. If you configure supplemental data for this Xero property, `suppliers` data from Xero could additionally include `BankAccountDetails`.

Reviewer: ```suggestion
Consider our Xero integration as an example. Codat's `suppliers` data type maps to Xero's [Contact](https://developer.xero.com/documentation/api/accounting/contacts) schema. Some of its properties (like `TaxNumber`) are included by us by default, whereas others (like `BankAccountDetails`) are not. If you configure supplemental data for this Xero property, `suppliers` data from Xero could additionally include `BankAccountDetails`.

```

> Consider our Xero integration as an example. Codat's `suppliers` data type uses Xero's [Contact](https://developer.xero.com/documentation/api/accounting/contacts) schema. Some of its properties (like `TaxNumber`) are mapped by us by default, whereas others (like `BankAccountDetails`) are not. If you configure supplemental data for this Xero property, `suppliers` data from Xero could additionally include `BankAccountDetails`.

Reviewer 2: ```suggestion
Consider our Xero integration. Xero's [Contact](https://developer.xero.com/documentation/api/accounting/contacts) schema maps to Codat's `supplier` data type. Some of its properties (like `TaxNumber`) *are* mapped, whereas others (like `BankAccountDetails`) are not. If you configure supplemental data for this Xero property, `suppliers` data from Xero could additionally include `BankAccountDetails`.

```

## Common uses of supplemental data

Integration-specific properties not included in our standard data models may still provide you with additional benefits, enriching the data relevant for your use case. For example, you can enhance our `invoices` data type with an invoice `URL`, a Xero-specific field that directs the user to a source document for an invoice hosted outside of Xero. For QBO, the `invoices` data type can be supplemented with `SalesTermRef`, providing you with sales terms associated with an invoice.

We have compiled a list of properties commonly used by our customers to enrich our standard data types. You can [review these in detail](/using-the-api/supplemental-data/usecases) to see how you can use supplemental data to your advantage.

## Support for supplemental data

We are rapidly expanding coverage across integrations and data types according to client demand. You can help us prioritize by leaving feedback on our [public roadmap](https://portal.productboard.com/codat/7-public-product-roadmap/tabs/46-accounting-api/submit-idea).

We currently cover the following integrations and data types:
 

> <iframe

Reviewer: Is it only commerce orders that are supported today? All our commerce data types were updated to support them in our OAS. I've reached out to Alessandro for confirmation. Depending on the answer we should update this google sheet or temporarily hit supplemental data from those data types for which it is not yet live.

Author: that's a good question, but not relevant to this document. the excel spreadsheet is open for them to edit, and I put reminders everywhere on their confluence to do that. and their internal docs don't say anything about other data types yet 
https://codatdocs.atlassian.net/wiki/spaces/IN/pages/2891317381/Supplemental+Data+-+Coverage

  src="https://docs.google.com/spreadsheets/d/e/2PACX-1vToBP6lQMT_MrB8L5e_61w2LrmpoJPAVhxCVqCuoSpWgb6ga2hUXZHlLSdCr9jY_He1b-uYaDAnH6DV/pubhtml?widget=true&amp;headers=false"
  frameborder="0"
  style={{ top: 0, left: 0, width: "100%", height: "400px" }}
></iframe>

## Configure supplemental data

> You'll need to specify what supplemental data should be passed in the response for each integration and data type pair you require. To do so, use the [Configure supplemental data](/codat-api#/operations/configure-supplemental-data) endpoint.

Reviewer: Thought: It would be useful for the free trial to have an example supplemental data setup for a data type so that a developer can see it in action. @mcclowes @pzaichkina. Would be a great for client enablement.

Author: created a story for this

```http
/integrations/{platformKey}/datatypes/{datatype}/supplementalDataConfig
```

Within the request body, `PlatformEndpoint` and `PlatformPropertyName` parameter values must match the integration's requirements exactly, including casing. Ensure you are familiar with the source data structure as Codat does not validate the supplemental data values against the integration provider.

```json title="Supplemental data configuration request body"
{

>     "supplementalDataConfig": {

Reviewer: Is the supplementalDataConfig needed? I thought it is just

```
"{ClientObjectName}": {
            "dataSource": "/{PlatformEndpoint}",
            "pullData": {
                "{ClientDefinedName}":"{PlatformPropertyName}",
                "{ClientDefinedName2}": "{PlatformPropertyName2}"
            },
            "pushData": {
                "{ClientDefinedName}":"{PlatformPropertyName}",
                "{ClientDefinedName2}": "{PlatformPropertyName2}"
            },
        }
```

Author: going off of this https://codatdocs.atlassian.net/wiki/spaces/~6331cf2e2eaaa5dcfa1422f6/pages/3042443333/Supplemental+Data+-+Draft+Docs#How-do-I-configure-Supplemental-Data%3F, but I will check

        "{ClientObjectName}": {
            "dataSource": "/{PlatformEndpoint}",
            "pullData": {
                "{ClientDefinedName}":"{PlatformPropertyName}",
                "{ClientDefinedName2}": "{PlatformPropertyName2}"
            },
            "pushData": {
                "{ClientDefinedName}":"{PlatformPropertyName}",
                "{ClientDefinedName2}": "{PlatformPropertyName2}"
            },
        }
    }
}
```
You can use dot notation to retrieve nested properties from within the supplemental data object. For example, maintain the following configuration to retrieve the supplier's `Name` value from Xero's `BrandingTheme` object with two properties, `BrandingThemeID` and `Name`, along with their `BankAccountDetails`.

```json title="Supplemental data configuration with dot notation"
{

>     "supplementalDataConfig": {

Reviewer: Might need updating depending on what the correct schema is.
